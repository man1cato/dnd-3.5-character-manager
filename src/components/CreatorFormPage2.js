import React from 'react';
import _ from 'lodash';
import { Field, ErrorMessage } from 'formik';

const validateProhibitedSchools = (selectedSchool) => (value) => {
	let error;
	if (selectedSchool === 'Divination') {
		if (value.length !== 1) {
			error = 'You must select one school whose magic use will be prohibited';
		}
	} else {
		if (selectedSchool !== 'Universal' && value.length !== 2) {
			error = 'You must select two schools whose magic use will be prohibited';
		}
	}
	return error;
}

const CreatorFormPage2 = ({values, selectedRace, classes, selectedClass, schools, handleChange, handleSelect, handleMultiSelect, setFieldValue, setFieldError}) => (
	<div>
		<div className="form__group form__content--flex">
			<h4>Choose Class:</h4>
			<Field name="jobClass" component="select" onChange={(e) => {handleChange(e); handleSelect(e)}}>
				{classes.map((jobClass) => (
					<option 
						value={jobClass.id} 
						key={jobClass.id}
					>
						{jobClass.name}
					</option>
				))}
			</Field>                                    
		</div>

		<div className="form__group form__content--flex">
			<h4>Hit Die:</h4>
			<div>{selectedClass.hitDie}</div>
		</div>

		<div className="form__group form__content--flex">
			<h4>Proficiencies:</h4>
			<div>{selectedClass.proficiencies}</div>
		</div>

		<div className="form__group form__content--flex">
			<h4>Choose Bonus Languages:</h4>
			<Field name="bonusLanguages" component="select" multiple onChange={(e) => {handleMultiSelect(e, setFieldValue)}}>
				{selectedRace.bonusLanguages.map((language) => (
					<option 
						value={language} 
						key={language}
					>
						{language}
					</option>
				))}
			</Field>                                    
		</div>

		<div className="form__group form__content--flex">
			<h4>Deity:</h4>
			<Field name="deity" placeholder="(Optional) Enter the name of a Deity" />                                  
		</div>
		
		{selectedClass.name === 'Wizard' && (
			<div className="form__group form__content--flex">
				<h4>Choose School:</h4>
				<Field name="school" component="select" onChange={(e) => {handleChange(e); handleSelect(e, setFieldValue, setFieldError)}}>
					{schools.map((school) => (
						<option 
							value={school} 
							key={school}
						>
							{school}
						</option>
					))}
				</Field>
			</div>
		)}

		{values.school !== 'Universal' && (
			<div className="form__group">
				<div className="form__content--flex">
					<h4>Choose Prohibited School(s):</h4>
					<Field 
						name="prohibitedSchools" 
						component="select" 
						multiple 
						validate={validateProhibitedSchools(values.school)}
						onChange={(e) => {handleMultiSelect(e, setFieldValue)}}
					>
						{_.difference(schools, [values.school, 'Universal', 'Divination']).map((school) => (
							<option 
								value={school} 
								key={`prohibited${school}`}
							>
								{school}
							</option>
						))}
					</Field>
				</div>
				<ErrorMessage name="prohibitedSchools" component="div"/>
			</div>
		)}
	</div>
);


export default CreatorFormPage2;
