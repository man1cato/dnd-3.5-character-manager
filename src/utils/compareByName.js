export default (a,b) => {
    const aName = a.name.toUpperCase();
    const bName = b.name.toUpperCase();
    if (aName < bName) { return -1 };
    if (aName > bName) { return 1 };
    return 0;
}