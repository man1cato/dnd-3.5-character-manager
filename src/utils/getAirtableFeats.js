import dotenv from 'dotenv'
dotenv.config()

import axios from 'axios'
import _ from 'lodash'

const apiKey = process.env.AIRTABLE_API_KEY
const baseUrl = 'https://api.airtable.com/v0/appK7TZeddGqjGUDL'

export default async () => {
	try {
		let offset
		let feats = []

		do {
			const res = await axios.get(`${baseUrl}/Feats?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
			offset = res.data.offset
			const featGroup = res.data.records.map((feat) =>  ({
				id: feat.id,
				name: feat.fields.Name,
				type: feat.fields.Type.join(", ") || null,
				description: feat.fields.Description || null
			}));
			
			feats = feats.concat(featGroup)
		} while (!!offset)
		
		return _.orderBy(feats, ['name'], ['asc'])
	} catch (e) {
		console.error(e)
	}
}