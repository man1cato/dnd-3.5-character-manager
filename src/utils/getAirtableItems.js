import dotenv from 'dotenv'
dotenv.config()

import axios from 'axios'
import _ from 'lodash'

const apiKey = process.env.AIRTABLE_API_KEY
const baseUrl = 'https://api.airtable.com/v0/appK7TZeddGqjGUDL'

export default async () => {
	try {
		let offset;
		let items = [];
		do {
			const res = await axios.get(`${baseUrl}/Items?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || ''))
			offset = res.data.offset
			const itemGroup = res.data.records.map((item) =>  ({
				id: item.id,
				name: item.fields.Name,
				value: item.fields["Value (gp)"] || '',
				weight: item.fields["Weight (lbs)"] || '',
				category: item.fields.Category || null,
				weaponType: item.fields["Weapon Type"] || null,
				encumbrance: item.fields.Encumbrance || null,
				damageType: item.fields["Damage Type"] ? item.fields["Damage Type"].join(" / ") : [],
				damageS: item.fields["Damage (S)"] || null,
				damageM: item.fields["Damage (M)"] || null,
				critical: item.fields.Critical || null,
				range: item.fields.Range || null,
				armorBonus: item.fields["Armor Bonus"] || null,
				maxDexBonus: item.fields["Max DEX Bonus"] || null
			}))
			
			items = items.concat(itemGroup)
		} while (!!offset)
		
		return _.orderBy(items, ['name'], ['asc'])
	} catch (e) {
		console.error(e)
	}
}