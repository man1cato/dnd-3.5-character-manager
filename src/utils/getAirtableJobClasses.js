import dotenv from 'dotenv'
dotenv.config()

import axios from 'axios'
import _ from 'lodash'

const apiKey = process.env.AIRTABLE_API_KEY
const baseUrl = 'https://api.airtable.com/v0/appK7TZeddGqjGUDL'

export default async () => {
	try {
		let offset
		let jobClasses = []
	
		do {
			const res = await axios.get(`${baseUrl}/Classes?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
			offset = res.data.offset
			const jobClassGroup = res.data.records.map((jobClass) =>  ({
				id: jobClass.id,
				name: jobClass.fields.Name,
				hitDie: jobClass.fields['Hit Die'],
				proficiencies: jobClass.fields['Proficiencies'] ? jobClass.fields['Proficiencies'].join(', ') : 'None',
				skills: jobClass.fields.Skills || []
			}))
			
			jobClasses = jobClasses.concat(jobClassGroup)
		} while (!!offset)
		
		return _.orderBy(jobClasses, ['name'], ['asc'])
	} catch (e) {
		console.error(e)
	}
}