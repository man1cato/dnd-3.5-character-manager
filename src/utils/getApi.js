import admin from 'firebase-admin'
import serviceAccount from '../../config/dnd-3-5-character-manager-key.json'

admin.initializeApp({
   credential: admin.credential.cert(serviceAccount),
   databaseURL: "https://dnd-3-5-character-manager.firebaseio.com"
 })
 
const database = admin.database()

export default async () => {
   try {
      const snapshot = await database.ref('api').once('value')
      return snapshot.val()
   } catch (e) {
      console.error(e)
   }
}
