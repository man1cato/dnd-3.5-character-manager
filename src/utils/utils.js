import _ from 'lodash'


export const convertTextToArray = (string) => _.compact(string.split(/\n/))

export const findItemById = (storeItemArray, itemId) => storeItemArray.find((item) => item.id === itemId)

export const apiObjectToArray = (obj) => _.orderBy(Object.keys(obj).map((id) => ({
	id,
	...obj[id]
})), ['name'], ['asc'])