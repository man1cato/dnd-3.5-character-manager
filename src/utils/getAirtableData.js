import axios from 'axios'
import _ from 'lodash'
import { convertTextToArray } from './utils' 
import dotenv from 'dotenv'
dotenv.config({ path: './config/dev.env' })

const apiKey = process.env.AIRTABLE_API_KEY
const baseUrl = 'https://api.airtable.com/v0/appK7TZeddGqjGUDL'


export const getFeats = async () => {
	try {
		let offset
		let feats = {}

		do {
			const res = await axios.get(`${baseUrl}/Feats?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
			offset = res.data.offset
			res.data.records.forEach((feat) => {
				feats[feat.id] = {
					name: feat.fields.Name,
					types: feat.fields.Types || null,
					description: feat.fields.Description || null,
					prerequisites: {
						level: feat.fields['Prereq. Level'] || null,
						class: feat.fields['Prereq. Class'] || null,
						feats: feat.fields['Prereq. Feats'] || null
					} || null
				}
			})			
		} while (!!offset)
		
		return feats
	} catch (e) {
		console.error(e)
	}
}


export const getItems = async () => {
	try {
		let offset
		let items = {}
		do {
			const res = await axios.get(`${baseUrl}/Items?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || ''))
			offset = res.data.offset
			res.data.records.forEach((item) => {
				items[item.id] = {
					name: item.fields.Name,
					value: item.fields["Value (gp)"] || '',
					weight: item.fields["Weight (lbs)"] || '',
					category: item.fields.Category || null,
					weaponType: item.fields["Weapon Type"] || null,
					encumbrance: item.fields.Encumbrance || null,
					damageType: item.fields["Damage Type"] ? item.fields["Damage Type"].join(" / ") : [],
					damageS: item.fields["Damage (S)"] || null,
					damageM: item.fields["Damage (M)"] || null,
					critical: item.fields.Critical || null,
					range: item.fields.Range || null,
					armorBonus: item.fields["Armor Bonus"] || null,
					maxDexBonus: item.fields["Max DEX Bonus"] || null
				}
			})
		} while (!!offset)
		
		return items
	} catch (e) {
		console.error(e)
	}
}


const getLevels = async (jobClassId) => {
	try {
		const levelRes = await axios.get(`${baseUrl}/Levels?filterByFormula={Class ID}="${jobClassId}"&api_key=${apiKey}`)
		
		const levelRecords = levelRes.data.records
		let levels = {}
		for (let i = 0; i < levelRecords.length; i++) {
			const levelRecord = levelRecords[i].fields 
			levels[levelRecord.Level] = {
				saves: {
					fortitude: levelRecord['Fortitude'],
					reflex: levelRecord['Reflex'],
					will: levelRecord['Will']
				},
				baseAttackBonuses: _.compact([ levelRecord['BAB 1'], levelRecord['BAB 2'], levelRecord['BAB 3'], levelRecord['BAB 4'] ]),
				xp: levelRecord['XP Required'],
				specialAbilities: !!levelRecord['Special Abilities'] ? levelRecord['Special Abilities'] : null,
				spellsPerDay:  _.compact([ 
					levelRecord['SPD 0'], levelRecord['SPD 1'], levelRecord['SPD 2'], levelRecord['SPD 3'], levelRecord['SPD 4'], 
					levelRecord['SPD 5'], levelRecord['SPD 6'], levelRecord['SPD 7'], levelRecord['SPD 8'], levelRecord['SPD 9'] 
				])
			}
		}
		return levels
	} catch (e) {
		console.error(e)
	}
}

export const getJobClasses = async () => {
	try {
		let offset
		let jobClasses = {}
	
		do {
			const classRes = await axios.get(`${baseUrl}/Classes?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
			offset = classRes.data.offset

			const jobClassGroup = classRes.data.records
			
			for (let i = 0; i < jobClassGroup.length; i++) {
				const jobClass = jobClassGroup[i] 
				jobClasses[jobClass.id] = {
					name: jobClass.fields.Name,
					hitDie: jobClass.fields['Hit Die'],
					proficiencies: jobClass.fields['Proficiencies'] || [],
					skills: jobClass.fields.Skills || [],
					levels: await getLevels(jobClass.id)
				}
			}
		} while (!!offset)
		
		return jobClasses
	} catch (e) {
		console.error(e)
	}
}


export const getRaces = async () => {
	try {
		let offset
		let races = {}
		do {
			const res = await axios.get(`${baseUrl}/Races?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
			offset = res.data.offset
			res.data.records.forEach((race) =>  {
				races[race.id] = {
					name: race.fields.Name,
					size: race.fields.Size,
					speed: race.fields['Speed (ft/rnd)'],
					racialBonuses: race.fields['Racial Bonuses'] ? convertTextToArray(race.fields['Racial Bonuses']) : null,
					defaultLanguages: race.fields['Languages (Default)'],
					bonusLanguages: race.fields['Languages (Bonus)'],
					abilityMods: {
						str: race.fields['STR Mod'],
						dex: race.fields['DEX Mod'],
						con: race.fields['CON Mod'],
						int: race.fields['INT Mod'],
						wis: race.fields['WIS Mod'],
						cha: race.fields['CHA Mod']
					},
					specialAbilities: race.fields['Special Abilities'] || [],
					favoredClass:  race.fields['Favored Class'] ? race.fields['Favored Class'][0] : null,
					iconUrl: race.fields['Icon URL']
				}
			})
			
		} while (!!offset)
		
		return races
	} catch (e) {
		console.error(e)
	}
}


export const getSkills = async () => {
   let offset
   let skills = {}
   do {
		const res = await axios.get(`${baseUrl}/Skills?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
		offset = res.data.offset
		res.data.records.forEach((skill) => {
			skills[skill.id] = {
				name: skill.fields.Name,
				keyAbility: skill.fields["Key Ability"] || null,
				description: skill.fields.Description || "",
				notes: skill.fields.Notes || ""
			}
		})
   } while (!!offset)
   
   return skills
}


export const getSpecialAbilities = async () => {
   let offset
   let specialAbilities = {}
   do {
		const res = await axios.get(`${baseUrl}/Special Abilities?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || ''))
		offset = res.data.offset
		res.data.records.forEach((specialAbility) => {
			specialAbilities[specialAbility.id] = {
				name: specialAbility.fields.Name,
				type: specialAbility.fields.Type ? specialAbility.fields.Type.join(", ") : null,
				description: specialAbility.fields.Description				
			}
		})
   } while (!!offset)
   
   return specialAbilities
}


export const getSpells =  async () => {
   let offset
   let spells = {}
   do {
		const res = await axios.get(`${baseUrl}/Spells?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '')) 
		offset = res.data.offset
		res.data.records.forEach((spell) => {
			spells[spell.id] = {
				name: spell.fields.Name,
				description: spell.fields.Description || "",
				school: spell.fields.School || null,
				components: spell.fields.Components.join(", ") || null,
				castingTime: spell.fields["Casting Time"] || null,
				range: spell.fields.Range || null,
				effectArea: spell.fields["Target, Effect, Area"] || null,
				duration: spell.fields.Duration || null,
				savingThrow: spell.fields["Saving Throw"] || null,
				spellResistance: spell.fields["Spell Resistance"] || null
			}
		})
   } while (!!offset)

   return spells
}