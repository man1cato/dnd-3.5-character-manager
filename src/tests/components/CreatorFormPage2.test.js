import React from 'react'
import { shallow } from 'enzyme'

import CreatorFormPage2 from '../../components/CreatorFormPage2'
import getApi from '../../utils/getFirebaseData'
import { apiObjectToArray } from '../../utils/utils'


const handleChange = jest.fn()
const handleSelect = jest.fn()
const setFieldValue = jest.fn()
const setFieldError = jest.fn()

let api, wrapper, props

beforeAll(async () => {
   api = await getApi()
   const races = apiObjectToArray(api.races)
   const jobClasses = apiObjectToArray(api.jobClasses)
   props = {
      jobClasses,
      selectedRace: races[0],
      selectedJobClass: jobClasses[0],
      handleChange,
      handleSelect, 
      setFieldValue,
      setFieldError
   }
})

beforeEach(() => {
   wrapper = shallow(<CreatorFormPage2 {...props} />)
})

test('should render CreatorFormPage2', () => {
	expect(wrapper).toMatchSnapshot()
})

