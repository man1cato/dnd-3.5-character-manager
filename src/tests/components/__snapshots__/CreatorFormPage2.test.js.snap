// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render CreatorFormPage2 1`] = `
<Fragment>
  <div
    className="form-group--35"
  >
    <h4>
      Alignment:
    </h4>
    <FormikConnect(FieldInner)
      className="select"
      component="select"
      name="alignment"
      onChange={[Function]}
    >
      <option
        key="alignment0"
        value="Lawful Good"
      >
        Lawful Good
      </option>
      <option
        key="alignment1"
        value="Neutral Good"
      >
        Neutral Good
      </option>
      <option
        key="alignment2"
        value="Chaotic Good"
      >
        Chaotic Good
      </option>
      <option
        key="alignment3"
        value="Lawful Neutral"
      >
        Lawful Neutral
      </option>
      <option
        key="alignment4"
        value="True Neutral"
      >
        True Neutral
      </option>
      <option
        key="alignment5"
        value="Chaotic Neutral"
      >
        Chaotic Neutral
      </option>
      <option
        key="alignment6"
        value="Lawful Evil"
      >
        Lawful Evil
      </option>
      <option
        key="alignment7"
        value="Neutral Evil"
      >
        Neutral Evil
      </option>
      <option
        key="alignment8"
        value="Chaotic Evil"
      >
        Chaotic Evil
      </option>
    </FormikConnect(FieldInner)>
  </div>
  <div
    className="form-group--35"
  >
    <h4>
      Job Class:
    </h4>
    <FormikConnect(FieldInner)
      className="select"
      component="select"
      name="jobClass"
      onChange={[Function]}
    >
      <option
        key="jobClass0"
        value="recbxccm8pp3fTmQw"
      >
        Barbarian
      </option>
      <option
        key="jobClass1"
        value="rec0GDUzwsKLYEyqM"
      >
        Bard
      </option>
      <option
        key="jobClass2"
        value="recawEdTMnIzdNnwR"
      >
        Cleric
      </option>
      <option
        key="jobClass3"
        value="recz6owTBNig43RWq"
      >
        Druid
      </option>
      <option
        key="jobClass4"
        value="reckjdfZ5f86z4Rtz"
      >
        Fighter
      </option>
      <option
        key="jobClass5"
        value="recVqqb9HsZRyEAC8"
      >
        Monk
      </option>
      <option
        key="jobClass6"
        value="recjjJTiERQ1ocuNh"
      >
        Paladin
      </option>
      <option
        key="jobClass7"
        value="recKnUPMQEmoCGcag"
      >
        Ranger
      </option>
      <option
        key="jobClass8"
        value="recQrfCKtksnnSsaZ"
      >
        Rogue
      </option>
      <option
        key="jobClass9"
        value="recniRmHZzspPDAFn"
      >
        Sorcerer 
      </option>
      <option
        key="jobClass10"
        value="recBCWC99ohXXKqwL"
      >
        Wizard
      </option>
    </FormikConnect(FieldInner)>
  </div>
  <div
    className="form-group--35 align-top"
  >
    <h4>
      Hit Die:
    </h4>
    <div>
      D12
    </div>
  </div>
  <div
    className="form-group--35 align-top"
  >
    <h4>
      Proficiencies:
    </h4>
    <div>
      Simple Weapons, Martial Weapons, Light Armor, Medium Armor, Shields
    </div>
  </div>
  <div
    className="form-group--35 align-top"
  >
    <h4>
      Bonus Languages:
    </h4>
    <FormikConnect(FieldInner)
      className="multi-select"
      component="select"
      multiple={true}
      name="bonusLanguages"
      onChange={[Function]}
    >
      <option
        key="Draconic"
        value="Draconic"
      >
        Draconic
      </option>
      <option
        key="Gnoll"
        value="Gnoll"
      >
        Gnoll
      </option>
      <option
        key="Gnome"
        value="Gnome"
      >
        Gnome
      </option>
      <option
        key="Goblin"
        value="Goblin"
      >
        Goblin
      </option>
      <option
        key="Orc"
        value="Orc"
      >
        Orc
      </option>
      <option
        key="Sylvan"
        value="Sylvan"
      >
        Sylvan
      </option>
    </FormikConnect(FieldInner)>
  </div>
  <div
    className="form-group--35"
  >
    <h4>
      Deity:
    </h4>
    <FormikConnect(FieldInner)
      className="text-input"
      name="deity"
      placeholder="(Optional) Enter the name of a Deity"
    />
  </div>
</Fragment>
`;
